File: src/uvfastapi/app/routes/superadmin.py
"""
routes/superadmin.py  —  SUPERADMIN-only endpoints.

All routes require:  role == SUPERADMIN

RAG / data:
  POST   /api/v1/superadmin/ingest                          — re-index the vector store

Admin management:
  GET    /api/v1/superadmin/admins                          — list all admins
  GET    /api/v1/superadmin/admins/pending                  — list pending (unapproved) admins
  PATCH  /api/v1/superadmin/admins/{admin_id}/approve       — approve a pending admin
  PATCH  /api/v1/superadmin/admins/{admin_id}/deactivate    — deactivate an active admin
  DELETE /api/v1/superadmin/admins/{admin_id}               — soft-delete an admin

User visibility (system-wide):
  GET    /api/v1/superadmin/users                           — list every user in the system
  GET    /api/v1/superadmin/users/{user_id}                 — a specific user's profile
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, HTTPException, Request, status

from uvfastapi.app.middleware.auth import SuperAdminDep
from uvfastapi.app.models import (
    AdminListResponse,
    ApproveAdminResponse,
    MessageResponse,
    PendingAdminResponse,
    UserListResponse,
    UserProfileResponse,
)
from uvfastapi.database.connection import get_pool
from uvfastapi.services import superadmin_service

router = APIRouter(
    prefix="/api/v1/superadmin",
    tags=["SuperAdmin"],
)


# ═════════════════════════════════════════════
#  RAG / Data ingestion
# ═════════════════════════════════════════════

@router.post(
    "/ingest",
    response_model=MessageResponse,
    summary="Re-index the vector store (superadmin only)",
)
async def ingest_data(request: Request, current_user: SuperAdminDep):
    """
    Triggers a full re-ingestion of source data into the vector store.
    The refreshed vectorstore is hot-swapped into app.state.
    """
    try:
        from uvfastapi.services.user_service import create_or_replace_vector_store
        from uvfastapi.rag_engine.orchestrator import get_vectorstore_for_retrieval

        create_or_replace_vector_store(request.app.state.embedding_function)
        request.app.state.vectorstore = get_vectorstore_for_retrieval(
            request.app.state.embedding_function
        )
        return MessageResponse(message="Vector store re-indexed successfully.")
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ingestion failed: {e}",
        )


# ═════════════════════════════════════════════
#  Admin management
# ═════════════════════════════════════════════

@router.get(
    "/admins",
    response_model=AdminListResponse,
    summary="List all admins (active and inactive)",
)
async def list_admins(current_user: SuperAdminDep) -> AdminListResponse:
    pool = await get_pool()
    return await superadmin_service.get_all_admins(pool)


@router.get(
    "/admins/pending",
    response_model=list[PendingAdminResponse],
    summary="List admins whose registration is pending approval",
)
async def list_pending_admins(current_user: SuperAdminDep) -> list[PendingAdminResponse]:
    pool = await get_pool()
    return await superadmin_service.get_pending_admins(pool)


@router.patch(
    "/admins/{admin_id}/approve",
    response_model=ApproveAdminResponse,
    summary="Approve a pending admin registration",
)
async def approve_admin(admin_id: uuid.UUID, current_user: SuperAdminDep) -> ApproveAdminResponse:
    pool = await get_pool()
    return await superadmin_service.approve_admin(pool, admin_id, current_user.id)


@router.patch(
    "/admins/{admin_id}/deactivate",
    response_model=MessageResponse,
    summary="Deactivate an active admin (blocks login without deleting)",
)
async def deactivate_admin(admin_id: uuid.UUID, current_user: SuperAdminDep) -> MessageResponse:
    pool = await get_pool()
    return await superadmin_service.deactivate_admin(pool, admin_id, current_user.id)


@router.delete(
    "/admins/{admin_id}",
    response_model=MessageResponse,
    summary="Soft-delete an admin account",
)
async def delete_admin(admin_id: uuid.UUID, current_user: SuperAdminDep) -> MessageResponse:
    pool = await get_pool()
    return await superadmin_service.delete_admin(pool, admin_id, current_user.id)


# ═════════════════════════════════════════════
#  User visibility (system-wide)
# ═════════════════════════════════════════════

@router.get(
    "/users",
    response_model=UserListResponse,
    summary="List every user in the system across all admins",
)
async def list_all_users(current_user: SuperAdminDep) -> UserListResponse:
    pool = await get_pool()
    return await superadmin_service.get_all_users(pool)


@router.get(
    "/users/{user_id}",
    response_model=UserProfileResponse,
    summary="Get a specific user's profile",
)
async def get_user(user_id: uuid.UUID, current_user: SuperAdminDep) -> UserProfileResponse:
    pool = await get_pool()
    return await superadmin_service.get_user_profile(pool, user_id)

File: frontend/js/superadmin.js
/**
 * superadmin.js — SuperAdmin-only operations.
 *
 * Exposes a single namespace: SuperAdmin
 *
 * Only called when the authenticated user's role is SUPERADMIN.
 * All functions hit /api/v1/superadmin/* endpoints.
 *
 * Responsibilities:
 *   Admin management:
 *     - List all admins (active + inactive).
 *     - List pending (unapproved) admins.
 *     - Approve a pending admin.
 *     - Deactivate an active admin.
 *     - Soft-delete an admin.
 *
 *   System-wide user visibility:
 *     - List every user across all admins.
 *     - Get a specific user's profile.
 *
 *   Data ingestion:
 *     - Trigger a full vector store re-index.
 *
 * None of these functions render UI directly — they return data or throw,
 * and the caller is responsible for updating the DOM.
 */

'use strict';

const SuperAdmin = (() => {

  // ── Admin management ──────────────────────────────────────────────────────

  /**
   * List all admins (active and inactive).
   * GET /api/v1/superadmin/admins
   * @returns {Promise<{ admins: object[], total: number }>}
   */
  async function listAdmins() {
    const res = await API.fetch('/api/v1/superadmin/admins');
    return _handleResponse(res, 'Failed to fetch admins.');
  }

  /**
   * List admins whose registration is pending approval.
   * GET /api/v1/superadmin/admins/pending
   * @returns {Promise<object[]>} Array of PendingAdminResponse
   */
  async function listPendingAdmins() {
    const res = await API.fetch('/api/v1/superadmin/admins/pending');
    return _handleResponse(res, 'Failed to fetch pending admins.');
  }

  /**
   * Approve a pending admin registration.
   * PATCH /api/v1/superadmin/admins/{admin_id}/approve
   * @param {string} adminId — UUID string
   * @returns {Promise<object>} ApproveAdminResponse
   */
  async function approveAdmin(adminId) {
    const res = await API.fetch(`/api/v1/superadmin/admins/${adminId}/approve`, {
      method: 'PATCH',
    });
    return _handleResponse(res, 'Failed to approve admin.');
  }

  /**
   * Deactivate an active admin (blocks login without deleting).
   * PATCH /api/v1/superadmin/admins/{admin_id}/deactivate
   * @param {string} adminId
   * @returns {Promise<{ message: string }>}
   */
  async function deactivateAdmin(adminId) {
    const res = await API.fetch(`/api/v1/superadmin/admins/${adminId}/deactivate`, {
      method: 'PATCH',
    });
    return _handleResponse(res, 'Failed to deactivate admin.');
  }

  /**
   * Soft-delete an admin account.
   * DELETE /api/v1/superadmin/admins/{admin_id}
   * @param {string} adminId
   * @returns {Promise<{ message: string }>}
   */
  async function deleteAdmin(adminId) {
    const res = await API.fetch(`/api/v1/superadmin/admins/${adminId}`, {
      method: 'DELETE',
    });
    return _handleResponse(res, 'Failed to delete admin.');
  }

  // ── System-wide user visibility ───────────────────────────────────────────

  /**
   * List every user in the system across all admins.
   * GET /api/v1/superadmin/users
   * @returns {Promise<{ users: object[], total: number }>}
   */
  async function listAllUsers() {
    const res = await API.fetch('/api/v1/superadmin/users');
    return _handleResponse(res, 'Failed to fetch users.');
  }

  /**
   * Get a specific user's profile (system-wide, not scoped to one admin).
   * GET /api/v1/superadmin/users/{user_id}
   * @param {string} userId — UUID string
   * @returns {Promise<object>} UserProfileResponse
   */
  async function getUser(userId) {
    const res = await API.fetch(`/api/v1/superadmin/users/${userId}`);
    return _handleResponse(res, 'Failed to fetch user profile.');
  }

  // ── Data ingestion ────────────────────────────────────────────────────────

  /**
   * Trigger a full re-ingestion of source data into the vector store.
   * POST /api/v1/superadmin/ingest
   * @returns {Promise<{ message: string }>}
   */
  async function reindexVectorStore() {
    const res = await API.fetch('/api/v1/superadmin/ingest', {
      method: 'POST',
    });
    return _handleResponse(res, 'Failed to trigger re-index.');
  }

  // ── Response helper ───────────────────────────────────────────────────────

  async function _handleResponse(res, fallback) {
    if (!res) throw new Error('No response — network error or session expired.');

    if (!res.ok) {
      let detail = fallback;
      try {
        const body = await res.json();
        detail = body.detail || fallback;
      } catch (_) { /* ignore */ }
      throw new Error(detail);
    }

    return res.json();
  }

  // ── Public interface ──────────────────────────────────────────────────────

  return {
    listAdmins,
    listPendingAdmins,
    approveAdmin,
    deactivateAdmin,
    deleteAdmin,
    listAllUsers,
    getUser,
    reindexVectorStore,
  };
})();